### PR TITLE
Fix scroller detection in angular directive

### LINKED ISSUE #246: Angular doesn't detect scroller when navigating through routes
It works only on first time when the page loads with headroom and it correctly detects the scroller.
Demo: http://codepen.io/norkunas/pen/WGbYxW?editors=1010

Just scroll down and you'll see that shadow is applied on the header, but clicking on "show subpage" and then scrolling down the shadow doesn't show up.

### PR DESCRIPTION
**Pull request by norkunas**
Fixes #246

### CODE REVIEW

**Comment by atsoy** [comment]
@norkunas probably I'm not alone with this issue. But I found #248 more useful in that case.

**Comment by norkunas** [comment]
@atsoy 👍

**Comment by WickyNilliams** [comment]
Excuse the delay! 

Can I ask, why the timeout? Is this still relevant if/when #248 gets merged?

**Comment by norkunas** [comment]
@WickyNilliams I don't remember now, but it fixed my issue. About #248 i think it's less optimal to watch for the scroller than to get it in a timeout

**Comment by tabish89** [comment]
@norkunas -- I am unable to get headroom.js to work with our angular environment; and, since you  seem to one the "resident angular implementers" of this library - I though I would reach out to you.  I have a note out to @WickyNilliams, but he says that he's not an expert at Angular.

The following has been done so far, but for the life of me cannot figure out why it is not working or what is wrong with my implementation:

1. Added headroom.js
2. Added angular.headroom.js
3. Added reference to index.html

```
  <script src="/js/libraries/angular.headroom.js"></script>
  <script src="/js/libraries/headroom.js"></script>
```

4. Added requirement to angularJS module

```
angular.module('PSApp', [
  'fcsa-number',
  'headroom',
  'typer',...
])
```

5. Added some basic CSS

```
.headroom {
    will-change: transform;
    transition: transform 200ms linear;
}

.headroom--pinned {
    transform: translateY(0%);
}

.headroom--unpinned {
    transform: translateY(-100%);
}
```

What am I doing wrong?  It just doesn't want to work... activate.  I scroll up/down and nothing happens...

Is this still a supported plugin (++ for, Angular).  Our site is developed using angular 1.2.x.

**Comment by norkunas** [comment]
@tabish89 I can't say what's wrong only with code without live example (html code included would shouw more about your usage), but I suggest to check the element when scrolling if the classes are applied at all. Maybe `scroller="#scroll-container"` is missing in your code if you use it as SPA and the scrolling container is not the `document.body`.

**Comment by tabish89** [comment]
@toddmotto 

My bad.  I don't want to be presumptuous about how much time you might have to help me, so let me begin by saying that I am extremely appreciative of your goodwill.  I'm a coding novice, so your patience is also much appreciated in advance.

In addition to the foregoing code snippets shared, here is an example of how I'm referencing headroom.js in HTML.  If you prefer, I can see how I can share our code base but that wouldn't let you set up a live example easily because there are a number of bespoke dependencies required in order to load the web application.

The 'codepen' most similar to how headroom.js is implemented on our end can be found [https://codepen.io/frankcastro/pen/KzLLpr?page=1&] here.  I obtained the library from NPM and have compared it to other versions on the web and:

1. **headroom.js** is the same
2. **angular-headroom.js** is the same; hwoever, I have seen another version out there (not merged) with a $timeout function

_I have not integrated a bespoke jQuery reference and was hoping to use the default settings built in Angular (I believe there is one)._

```

<headroom tolerance="0" offset="0" classes="{pinned:'headroom-pinned',unpinned:'headroom-unpinned',initial:'headroom'}">
TEST HEADROOM

  <md-toolbar class="md-primary md-hue-3 header" layout="row" layout-align="start center" hide-sm hide-xs>

    <a ui-sref="site.main"><img id="headerLogo" src="/img/logos/logodo-08-whitebk-beta.png"></a>

    ...

    <md-button class="md-icon-button" ng-click="toggleSideNav()" aria-label="SideNav">
       <md-icon md-svg-icon="menu" style="fill: white;"></md-icon>
    </md-button>

  </md-toolbar>

</headroom>

```

I checked CONSOLE (Google Chrome) to see whether the classes are being applied; and, they seem to be there.

```

.headroom-pinned {                global.css:47
   transform: translateY(-100%);
}
_Appears when scrolling up


.headroom-unpinned {            global.css:43
   transform: translateY(0%);
}
Appears when scrolling down


.headroom{                               global.css:38
   will-change: transform;
   transition: transform 200ms
       [x] linear;
}

```

Major headscratcher...

**Comment by norkunas** [comment]
I'm using headroom directly on md-toolbar directive instead of wrapping it. Please try:
```
<md-toolbar headroom tolerance="0" offset="0" classes="{pinned:'headroom-pinned',unpinned:'headroom-unpinned',initial:'headroom'}" class="md-primary md-hue-3 header" layout="row" layout-align="start center" hide-sm hide-xs>
  <a ui-sref="site.main"><img id="headerLogo" src="/img/logos/logodo-08-whitebk-beta.png"></a>
    ...
  <md-button class="md-icon-button" ng-click="toggleSideNav()" aria-label="SideNav">
    <md-icon md-svg-icon="menu" style="fill: white;"></md-icon>
  </md-button>
</md-toolbar>
```

**Comment by tabish89** [comment]
Just tried your recommendation, but still not working.  :-(  Here is the whole HTML to make sure nothing is possibly conflicting with the headroom.js code:

```

<md-toolbar headroom tolerance="205" offset="5" classes="{pinned:'headroom-pinned',unpinned:'headroom-unpinned',initial:'headroom'}" class="md-primary md-hue-3 header" layout="row" layout-align="start center" hide-sm hide-xs>
  <a ui-sref="site.main"><img id="headerLogo" src="/img/logos/logodo-08-whitebk-beta.png"></a>

  <span flex></span>

  <md-button style="text-transform: none;" ng-if="!isLoggedIn">
    <a style="color: white; font-size: 14px; font-family: 'Lato', Fallback, sans-serif" ui-sref="authentication.login">Sign In</a>
  </md-button>

  <md-button style="text-transform: none;">
    <a style="color: white; font-size: 14px; font-family: 'Lato', Fallback, sans-serif" ui-sref="site.product.features">Features</a>
  </md-button>

  <md-button ng-if="!isLoggedIn" style="text-transform: none;">
    <a style="color: white; font-size: 14px; font-family: 'Lato', Fallback, sans-serif" ui-sref="site.pricing">Pricing</a>
  </md-button>

  <md-button ng-if="isLoggedIn" style="text-transform: none;">
    <a style="color: white; font-size: 14px; font-family: 'Lato', Fallback, sans-serif" ui-sref="console.subscription">Subscription</a>
  </md-button>

  <md-button style="text-transform: none;">
    <a style="color: white; font-size: 14px; font-family: 'Lato', Fallback, sans-serif" ui-sref="site.splash.company">Company</a>
  </md-button>

  <md-menu ng-controller="SiteCtrl.Menu as ctrl">
    <md-button style="text-transform: none" ng-click="$mdMenu.open($event)">
      <a style="color: white; font-size: 14px; font-family: 'Lato', Fallback, sans-serif">FAQ</a>
    </md-button>
    <md-menu-content class="light-gray-background" width="3">
      <md-menu-item ng-repeat="item in ['General', 'Business', 'Technology']">
        <md-button ng-click="ctrl.selectItem($index, false)">
          <div layout="row" flex>
            <p flex>{{item}}</p>
          </div>
        </md-button>
      </md-menu-item>
    </md-menu-content>
  </md-menu>

  <md-button class="md-icon-button" ng-click="toggleSideNav()" aria-label="SideNav">
    <md-icon md-svg-icon="menu" style="fill: white;"></md-icon>
  </md-button>
</md-toolbar>

```

I saw some other variations on this on the web, where someone set it up as a FACTORY 'Directive' (I've tried this too, but not working)...

```
  .directive('headroom', function() {
    return {
      restrict: 'E',
      replace: true,
      transclude: true,
      template: '<div ng-transclude></div>',
      link: function(scope, element, attrs) {
        var headroom, options;

        var setupOptions = function() {
          if (options)
            return;

          options = [];
          angular.forEach(Headroom.options, function(value, key) {
            if (key == 'classes' && attrs[key])
              options[key] = angular.fromJson(attrs[key]);
            else
              options[key] = attrs[key] || Headroom.options[key];
          });

          if (attrs.onPin) {
            options.onPin = function() {
              scope.$apply(attrs.onPin);
            };
          }
          if (attrs.onUnpin) {
            options.onUnpin = function() {
              scope.$apply(attrs.onUnpin);
            };
          }
          if (attrs.onTop) {
            options.onTop = function() {
              scope.$apply(attrs.onTop);
            };
          }
          if (attrs.onNotTop) {
            options.onNotTop = function() {
              scope.$apply(attrs.onNotTop);
            };
          }
        };

        var initialize = function() {
          setupOptions();
          headroom = new Headroom(element[0], options);
          headroom.init();
        };

        var destroy = function() {
          if (headroom)
            headroom.destroy();
        };

        attrs.$observe('disabled', function(value) {
          if (angular.isUndefined(value))
            return;

          if (value === true)
            destroy();
          else
            initialize();
        });

        scope.$on('$destroy', function() {
          destroy();
        });
      }
    };
  });

```

**Comment by norkunas** [comment]
I think you need to set the `scroller="md-content"` or to some other
selector which is your content with scrolling.

2017-08-24 16:16 "Tabish Rizvi" <notifications@github.com> rašė:

**Comment by tabish89** [comment]
Tried it both ways... simply doesn't want to work.

```
<headroom scroller="md-content" tolerance="205" offset="5" classes="{pinned:'headroom-pinned',unpinned:'headroom-unpinned',initial:'headroom'}">
TEST HEADROOM

  <md-toolbar class="md-primary md-hue-3 header" layout="row" layout-align="start center" hide-sm hide-xs>

...

</md-toolbar>
</headroom>

or:

```
<md-toolbar headroom scroller="md-content" tolerance="205" offset="5" classes="{pinned:'headroom-pinned',unpinned:'headroom-unpinned',initial:'headroom'}" class="md-primary md-hue-3 header" layout="row" layout-align="start center" hide-sm hide-xs>
```

**Comment by tabish89** [comment]
Think I might have figured out "WHY" it is not working.  Here is a [https://codepen.io/tabish89/pen/eEjdrj] CodePen with a lite-working version of the code.

You will notice in the HTML that there is a `md-menu` embedded in my 'md-toolbar', which seems to be conflicting with the headroom code.  When it is disabled, `headroom` works.  When enabled, `headroom` breaks.

Now, I am only able to get `headroom` to work on codepen and not my actual server, even when I disable `md-menu`. :-( . 

That means it might have something to do with the use of an angular `ui-router` and how the HTML is built using partials.  Have you heard of any such issue?

```
The dependencies for the codepen are:

1. Actual angular.js library installed locally on my server (https://pelotonstreet.com/js/libraries/angular.min.js)

2. Headroom.js from (https://npmcdn.com/headroom.js), which is the same one installed locally.

3. jQuery from (https://cdnjs.cloudflare.com/ajax/libs/jquery/2.2.2/jquery.min.js)

4. Angular-headroom.js from another codepen (https://codepen.io/frankcastro/pen/VaOOpp), which is identical to the one installed locally.

```

Am I missing a reference or something with the jquery library?

**Comment by tabish89** [comment]
@norkunas After much trial-and-error, I managed to get it work.  Thanks so much for your help.

**Comment by norkunas** [comment]
@tabish89 glad you resolved it :+1:

**Comment by WickyNilliams** [comment]
@tabish89 glad you got it resolved :) It might be worth explaining here how you fixed it, should anyone else have the problem in future...

Also, a word on etiquette (since you said you're a novice) - it's not cool to spam github issues with requests for help. You commented on 4 separate issues, which is totally unnecessary. For personal help requests (where there is clearly not a bug), you would be better served posting to StackOverflow where there is a veritable army of people happy to help diagnose problems.

And thanks @norkunas for stepping up to assist

**Comment by tabish89** [comment]
@WickyNilliams You know you spend a lot of time telling people about etiquette, but perhaps you should spend that time updating the poor documentation that accompanies your library.  So call it what you want.

You disappear for months on end, but yet log on to tell someone about etiquette or how they're "being rude."  Frankly, just reading your comments and responses irk me - almost hypocritical (yes, I'm calling you out). You really think that you're God's gift to humanity?!?  I understand that it is open source code and that you developed headroom.js on your own time free of charge (tell me how big of a violin you want me to play), but you should still take ownership of your thing otherwise remove it from the web/domain all together.  I think there are several people here who would directionally agree with me.

I tried StackOverflow, CodePen, and a few others -- and most of the interaction on those sites were over 2 years old, so that tells me people have generally given up.  I did however notice some more recent comments on Github; and, decided to reach out to others in the community and see if we could solve the issue (because of your subpar and lacking documentation) together.  That's when @norkunas stepped up to plate and kindly pointed me in the right direction.

_"Chickity-check yo self before you wreck yo self"_

Mike drop...

**Comment by tabish89** [comment]
For the rest of the community, here is codepen to see how I fixed the issue: [codepen](https://codepen.io/tabish89/pen/eEjdrj)

Couple of notes:

1.  Take a look at the JS dependencies by clicking on the "gear" icon in the JS window.  You will notice that there is one external dependency to another [codepen created by Frank Castro](https://codepen.io/frankcastro/pen/VaOOpp), which is essentially the anguar.headroom.js wrapper.  I used this external link instead of the one of our server, because all of the development related to headroom.js has been done on a branch that does not have an externally facing IP.  Our production server has not been synched up with our dev branches, since there is still a lot of work in progress.

2. Some salient differences in the CSS.  What is provided in the documentation would not work, but this seems to have worked.  Not sure why... but, it's above my pay grade.

3. In some cases, md-menu in Angular can cause some conflicts -- so, it is disabled in the codepen.

**Comment by wesleycho** [comment]
@tabish89 to be fair, GitHub is not the place to ask the questions you have been asking. No repository owner/maintainer should be forced to have to respond to everyone by a certain time - you have no right to make that demand on someone else’s time whatsoever. You also can fork anything put in the open if you so wish - nobody is stopping you from doing so.

Comments like that frankly are why a lot of good maintainers don’t want to work on open source software anymore. Offloading personal frustrations over using someone’s library to increase other people’s stress is not ok. Unwarranted disrespect hurts every user.

Filing an issue over insufficient documentation on how to do certain things would be ok - it is important to remember that the informal contract is that maintainers give their time to improve libraries, almost always out of charity, but there is an equal burden for users to be good actors & foster respectful collaboration.

**Comment by tabish89** [comment]
@wesleycho Fair.  I see your point.  It's all in the delivery.  I had tried to be very careful in not demanding any feedback or response by a certain time -- and, apologize if my comments were misconstrued as such.

**Comment by WickyNilliams** [comment]
@tabish89 which exact parts of my comment elicited that response? It's absolutely baffling that you think that is a reasonable way to respond. I had, up until then, been nothing but polite and courteous. That may not continue. And worse yet, you seem proud of how you behave, given that trite "Mike drop" sign off (it's _mic_ btw). It is nothing to be proud of, it is unequivocally shameful behaviour. 

You are, by your own admission, a novice. Therefore you are likely inexperienced in dealing with open source projects. So I tried to explain (politely, I might add) some things for your benefit, and you come back with insults and an abrasive attitude.

So let me break it down, so that there is no room for further misunderstanding...

> I am extremely appreciative of your goodwill. I'm a coding novice, so your patience is also much appreciated in advance.

Your goodwill, patience, and appreciation didn't last long, did it? If you're inexperienced, how about listening, and taking the advice of someone who is not?

> You disappear for months on end

Because I have a life, and I can choose to work on _my_ projects whenever I want.

> yet log on to tell someone about etiquette

Because you obviously do not understand the etiquette, and it absolutely needs explaining. How you acted initially suggested this was the case, and how you followed up proved it beyond a shadow of a doubt.

> You really think that you're God's gift to humanity?!?

Ummm... no? I merely expect to be spoken to, and treated with, basic decency. This relationship is not mutually beneficial. You get free software, I get nothing but demands on my time, and occasionally having to ensure people like you (though I must say, you are by far the worst I have come across. Congratulations, I guess?). So how about being a little more cordial in future, seeing as it is you that is getting all the benefit out of this relationship.

> I tried StackOverflow, CodePen, and a few others

But did you _post_ to stack overflow? That's what I was suggesting is the best course for future. The reason being I know that 90% of issues raised here are not bugs but implementation issues. Or to be curt: it is your problem, not mine. All you do is distract from real issues by posting personal help requests here.

> and most of the interaction on those sites were over 2 years old, so that tells me people have generally given up.

Or, the more likely scenario, that most people are not having issues and are able to use the library without issue. This library is used by a huge number of websites (>2k of the top 1m sites!), all without issue. Did you stop to consider that _your_ code might be the problem?

> I understand that it is open source code and that you developed headroom.js on your own time free of charge (tell me how big of a violin you want me to play), but you should still take ownership of your thing otherwise remove it from the web/domain all together.

I am under no such obligation. Nor is any other open source maintainer. Have you read the MIT license? It says as much: "This software is provided as is". You are free to fork this project, write a competing project from scratch, or make pull requests if something is lacking. I encourage you to do so!

> I think there are several people here who would directionally agree with me.

I am struggling to imagine a single person agreeing with you. In fact, I expect people would be unanimously appalled by your attitude. 

> I did however notice some more recent comments on Github; and, decided to reach out to others in the community.

You did not reach out to the community, as this is not a community, nor is it a place to make personal help requests - it is a bug tracker. You spammed several issues here instead of replying to my offer of help on twitter (and _continued_ to spam issues after I asked you not to via twitter). Even creating a brand new issue would have been preferable, since at least then only I'm getting notifications, rather than the 10 or so people in all the issues you commented on.

> (because of your subpar and lacking documentation) 

Where was it lacking, and in what way? I assume you're working on a PR to improve the docs? Right?

> I see your point. It's all in the delivery. I had tried to be very careful in not demanding any feedback or response by a certain time -- and, apologize if my comments were misconstrued as such.

Nope, you've completely missed the point. The point is you're acting rude, insulting, and entitled. _That_ is what you should apologise for, and what you should avoid in future.

**TL;DR:** I think it is you, sir, that should "check yourself", because you are acting obnoxious and entitled, which is not how you be a good member of the open source community. I hope your future interactions with maintainers will be less hostile, and wish you the best.

**Comment by tabish89** [comment]
@WickyNilliams 

You’re correct... I did not see your response to my Twitter request.  In my ripe old age of 42, I'm NOT all that plugged into Twitter and Instagram... but, that's not an excuse.

Given that outreach on your part, I am out of line.  Global apology (more, for wasting your time in having to write such a length response).  When I received your response this morning, I was under the (wrong) impression that you were accosting me (like you have to a few others here) from the get go - which is what incited my rude response back to you.  Like I said... out of line.

Lesson learned.

**Comment by WickyNilliams** [comment]
Thank you for being mature and polite in this response.

But! It is out of line whether I had replied on twitter or not. Read my original comment again. There is nothing there that should provoke that response, it started with "glad you got it resolved :)" - clearly friendly in tone. The twitter element has very little bearing.

I reject the notion I have accosted anyone. Accosting is a very strong word. People have brought attitude and a sense of entitlement here time and time again. It is not accosting someone to call it out. Where I have been wrong I have apologised (communicating in text can be difficult). However that attitude and rudeness is common in OSS. If you ever release a project that gains some success you will see it first hand. Speak to any maintainer and they will tell you the same. It is draining and demoralising. For the most part I now avoid open source because of it. Which is why, as you observed, I only drop in every couple of months, and only because I feel some guilt for neglecting my projects.

**Comment by WickyNilliams** [comment]
BTW, did you have any suggestions to improve the docs? Happy to update them if there's any glaring omissions

**Comment by tabish89** [comment]
FWIW, at for the Angular implementation since it can be a bit of a crapshoot, a working codepen based that shows live/working example would be helpful.   You might consider cleaning up by forking what I used to troubleshoot the issue:
